local-volume/provisioner/pkg/deleter/deleter.go

```diff
@@ -79,27 +73,41 @@ func (d *Deleter) cleanupPV(pv *v1.PersistentVolume) error {
 		return fmt.Errorf("Unknown storage class name %v", pv.Spec.StorageClassName)
 	}
 
+	mountPath, err := common.GetPathFromPV(pv, config)
+	if err != nil {
+		return err
+	}
+
 	// TODO: Get volType from PV.
 	volType := common.VolumeTypeFile
+	isblk, _ := d.VolUtil.IsBlock(mountPath)
+	if isblk {
+		volType = common.VolumeTypeBlock
+	}
+
 	switch volType {
 	case common.VolumeTypeFile:
-		return d.cleanupFileVolume(pv, config)
+		return d.deleteFilePV(pv, mountPath, config)
 	case common.VolumeTypeBlock:
-		return fmt.Errorf("Not yet implemented")
+		return d.blockCleaner.DeleteBlockPV(pv, mountPath, config)
 	default:
 		return fmt.Errorf("Unexpected volume type %q for deleting path %q", volType, pv.Spec.Local.Path)
 	}
 }
 
-func (d *Deleter) cleanupFileVolume(pv *v1.PersistentVolume, config common.MountConfig) error {
-	specPath := pv.Spec.Local.Path
-	relativePath, err := filepath.Rel(config.HostDir, specPath)
-	if err != nil {
-		return fmt.Errorf("Could not get relative path: %v", err)
+func (d *Deleter) deleteFilePV(pv *v1.PersistentVolume, mountPath string, config common.MountConfig) error {
+	glog.Infof("Deleting PV FILE volume %q contents at hostpath %q, mountpath %q", pv.Name, pv.Spec.Local.Path,
+		mountPath)
+	if err := d.VolUtil.DeleteContents(mountPath); err != nil {
+		return err
 	}
 
-	mountPath := filepath.Join(config.MountDir, relativePath)
-
-	glog.Infof("Deleting PV %q contents at hostpath %q, mountpath %q", pv.Name, specPath, mountPath)
-	return d.VolUtil.DeleteContents(mountPath)
+	// Remove API object
```

Comment:
Maybe file delete should also be asynchronous, so you have the same handling for both.

Reply:
OK

Reply:
done